src/libcore/ops/drop.rs

```diff
@@ -34,52 +34,95 @@
-/// `Outer`, then for `Inner`. Therefore, `main` prints `Dropping Outer!` and
-/// then `Dropping Inner!`.
+/// # impl Drop for HasDrop {
+/// #     fn drop(&mut self) {
+/// #         println!("Dropping!");
+/// #     }
+/// # }
+/// fn main() {
+///     let _x = HasDrop;
 ///
+///     let mut x = _x;
+///     Drop::drop(&mut x);
+/// }
 /// ```
-/// struct Inner;
-/// struct Outer(Inner);
 ///
-/// impl Drop for Inner {
+/// As you can see, our custom implementation of `Drop` will be called at the end of `main`.
+///
+/// ## You cannot call `Drop::drop` yourself
+///
+/// Because the compiler automatically calls `Drop::drop`, you cannot call it yourself. This
+/// would lead to "double drop", where `Drop::drop` is called twice on the same value. This
+/// can lead to things like "double frees".
+///
+/// In other words, if you tried to write the code in the previous example with the explicit
+/// call to `Drop::drop`, you'd get a compiler error.
+///
+/// If you'd like to call `drop` yourself, there is something you can do: call [`std::mem::drop`].
+/// This function will drop its argument. For more, see its documentation.
+///
+/// [`std::mem::drop`]: ../../std/mem/fn.drop.html
+///
+/// ## `Drop` is recursive
+///
+/// If your type is something like a `struct` or `enum` that is an aggregate of other types, then
+/// Rust will call `Drop::drop` on the type first, and then recursively on everything it contains.
+/// For example:
+///
+/// ```rust
+/// struct HasDrop;
+///
+/// impl Drop for HasDrop {
 ///     fn drop(&mut self) {
-///         println!("Dropping Inner!");
+///         println!("Dropping HasDrop!");
 ///     }
 /// }
 ///
-/// impl Drop for Outer {
+/// struct HasTwoDrops {
+///     one: HasDrop,
+///     two: HasDrop,
+/// }
+///
+/// impl Drop for HasTwoDrops {
 ///     fn drop(&mut self) {
-///         println!("Dropping Outer!");
+///         println!("Dropping HasTwoDrops!");
 ///     }
 /// }
 ///
 /// fn main() {
-///     let _x = Outer(Inner);
+///     let _x = HasTwoDrops { one: HasDrop, two: HasDrop };
 /// }
 /// ```
 ///
-/// ## Variables are dropped in reverse order of declaration
-///
-/// `_first` is declared first and `_second` is declared second, so `main` will
-/// print `Declared second!` and then `Declared first!`.
+/// This will print
 ///
+/// ```text
+/// Dropping HasTwoDrops!
+/// Dropping HasDrop!
+/// Dropping HasDrop!
 /// ```
-/// struct PrintOnDrop(&'static str);
 ///
-/// impl Drop for PrintOnDrop {
-///     fn drop(&mut self) {
-///         println!("{}", self.0);
-///     }
-/// }
+/// Similarly, a slice will drop each element in order.
 ///
-/// fn main() {
-///     let _first = PrintOnDrop("Declared first!");
-///     let _second = PrintOnDrop("Declared second!");
-/// }
-/// ```
+/// Which of our two `HasDrop` drops first, though? It's the same order that
+/// they're declared: first `one`, then `two`. If you'd like to try this
```

Comment:
not true for stack slots. [play link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&code=struct%20Test(%26%27static%20str)%3B%0A%0Aimpl%20Drop%20for%20Test%20%7B%0A%20%20%20%20fn%20drop(%26mut%20self)%20%7B%0A%20%20%20%20%20%20%20%20println!(%22on%20drop%3A%20%7B%7D%22%2C%20self.0)%3B%0A%20%20%20%20%7D%0A%7D%0A%0Afn%20main()%20%7B%0A%20%20%20%20let%20x%20%3D%20Test(%22x%22)%3B%0A%20%20%20%20let%20y%20%3D%20Test(%22y%22)%3B%0A%20%20%20%20println!(%22Hello%2C%20world!%22)%3B%0A%7D)

Reply:
sure, this is not a stack slot though.